src/chunkserver/bgjobs.cc

```diff
@@ -42,608 +30,358 @@
-	zassert(pthread_attr_destroy(&thattr));
-	return jp;
 }
 
-uint32_t job_pool_jobs_count(void *jpool) {
-	TRACETHIS();
-	jobpool* jp = (jobpool*)jpool;
-	return jp->jobsQueue->elements();
+void JobPool::disableJob(uint32_t jobId) {
+	std::unique_lock jobsUniqueLock(jobsMutex, std::defer_lock);
+	auto jobIterator = jobHash.find(jobId);
+	if (jobIterator != jobHash.end()) {
+		jobsUniqueLock.lock();
+		if (jobIterator->second->state == JobPool::State::Enabled) {
+			jobIterator->second->state = JobPool::State::Disabled;
+		}
+		jobsUniqueLock.unlock();
+	}
 }
 
-void job_pool_disable_and_change_callback_all(void *jpool,void (*callback)(uint8_t status,void *extra)) {
-	TRACETHIS();
-	jobpool* jp = (jobpool*)jpool;
-	uint32_t jhpos;
-	job *jptr;
+void JobPool::checkJobs() {
+	uint32_t jobId;
+	uint8_t status;
 
-	std::lock_guard jobsLockGuard(jp->jobsMutex);
+	receiveStatus(jobId, status);
 
-	for (jhpos = 0 ; jhpos<JHASHSIZE ; jhpos++) {
-		for (jptr = jp->jobhash[jhpos] ; jptr ; jptr=jptr->next) {
-			if (jptr->jstate==JSTATE_ENABLED) {
-				jptr->jstate=JSTATE_DISABLED;
-			}
-			jptr->callback=callback;
+	auto jobIterator = jobHash.find(jobId);
+	if (jobIterator != jobHash.end()) {
+		auto callback = jobIterator->second->callback;
+		if (callback) {
+			callback(status, jobIterator->second->extra);
 		}
+		jobHash.erase(jobIterator);
 	}
 }
 
-void job_pool_disable_job(void *jpool,uint32_t jobid) {
-	TRACETHIS();
-	jobpool* jp = (jobpool*)jpool;
-	uint32_t jhpos = JHASHPOS(jobid);
-	job *jptr;
+void JobPool::changeCallback(uint32_t jobId, JobCallback callback, void *extra) {
+	std::unique_lock lock(jobsMutex);
+	auto jobIterator = jobHash.find(jobId);
+	if (jobIterator != jobHash.end()) {
+		jobIterator->second->callback = std::move(callback);
+		jobIterator->second->extra = extra;
+	}
+}
 
-	std::unique_lock jobsUniqueLock(jp->jobsMutex, std::defer_lock);
+void JobPool::workerThread() {
+	static std::atomic_uint16_t workersCounter(0);
+	std::string threadName = "jobWorker " + std::to_string(workersCounter++);
+	pthread_setname_np(pthread_self(), threadName.c_str());
 
-	for (jptr = jp->jobhash[jhpos] ; jptr ; jptr=jptr->next) {
-		if (jptr->jobid==jobid) {
-			jobsUniqueLock.lock();
-			if (jptr->jstate==JSTATE_ENABLED) {
-				jptr->jstate=JSTATE_DISABLED;
-			}
+	uint32_t jobId;
+	uint32_t operation;
+	uint8_t *jobPtrArg;
+	State jobState;
+	uint8_t status = SAUNAFS_STATUS_OK;
+
+	std::unique_lock jobsUniqueLock(jobsMutex, std::defer_lock);
+
+	while (true) {
+		jobsQueue->get(&jobId, &operation, &jobPtrArg, nullptr);
+		Job *job = reinterpret_cast<Job *>(jobPtrArg);
+
+		jobsUniqueLock.lock();
+		if (job == kInvalidJob) {
+			jobState = State::Disabled;
+		}
+		else {
+			jobState = job->state;
+			if (job->state == State::Enabled) { job->state = State::InProgress; }
+		}
+
+		if (operation == ChunkOperation::Exit) { break; }
+
+		if (jobState == State::Disabled) {
+			status = SAUNAFS_ERROR_NOTDONE;
 			jobsUniqueLock.unlock();
+			continue;
```

Comment:
Previous version of the code was sending even the statuses of the disabled jobs, which is `SAUNAFS_ERROR_NOTDONE`, is this happening here?

P.D: I've seen a commit solving a deadlock in these lines. Maybe this code solves it as well?:
```suggestion
		if (jobState == State::Disabled) {
			status = SAUNAFS_ERROR_NOTDONE;
			jobsUniqueLock.unlock();
			sendStatus(jobId, status);
			continue;
```

Reply:
You're right. The previous version also sent `status` for jobs having `Disabled` status. It means that `Disabled` jobs are processed the same as jobs in other states. My reasoning behind this change was that `Disabled` jobs should not be processed. That's why I removed the line sending `status` for jobs in `Disabled` state.

Maybe, my assumption was wrong and the correct solution is to keep the previous behavior by following your suggestion. I noticed this after adding the unit test for `disableJobs()` and at this moment, I decided to change it. 

Considering that I'm not sure whether my assumption is correct, I will add your suggestion to keep the original behavior of the system and I will update the unit test for `DisableJob`.

Thank you for pointing out @dmga44  :+1: